### DISCUSSION TITLE
Frictional contact with locking correction has bad convergence

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#20856] [Q&A General] [answered]
Hello,

I'm running a simulation where frictional contact is applied between two bodies. Without locking correction, the results are incorrect (I have an Abaqus simulation for comparison) and the convergence is good. When I turn locking correction on, the results are correct but the convergence takes a huge hit. The solve time increases from 3 to 10 minutes. I already played with a lot of options like the ones in the contact-block, scaling and PETsc preconditioning but I can't get it to converge properly.

Any ideas to fix this behaviour?

**Comment by GiudGiud** [comment]
Hello

Do you know what methods Abaqus is using to solve that same problem?

Guillaume

**Reply by Flolaffel** [reply]
Abaqus uses so called hybrid elements that introduce an additional DOF to compute the pressure stress in the elements to model (nearly) incompressible materials (I'm using a rubber part in my contact).

**Reply by GiudGiud** [reply]
how does it solve the nonlinear problem? like the numerical method (newton? pjfnk? and the linear problems at each step if one of those)

**Reply by Flolaffel** [reply]
Abaqus used the Newton method with a direct solver based on Gauss elimination for the linear equations.

**Reply by recuero** [reply]
We don't do many hyperelastic material simulations: Do you know it's friction that triggers this behavior? I.e. is it harder for the solver to converge with frictionless contact?

**Reply by Flolaffel** [reply]
Frictionless contact convergence is a lot better than frictional. Same simulation with frictionless formulation only takes 3 minutes instead of 10 and needs a lot less iterations.

**Reply by recuero** [reply]
Yeah, but I was mostly referring to the interaction with the correction to volumetric locking.

**Reply by recuero** [reply]
Also, could you paste your contact block?

**Reply by Flolaffel** [reply]
I've run some tests and here are the numbers I obtained:

- frictionless, no locking correction: 83 nonlinear iterations, 229 linear iterations
- frictionless, locking correction: 83 nonlinear iteration, 285 linear iterations
- frictional, no locking correction: 111 nonlinear iterations, 2961 linear iterations
- frictional, locking correction: 191 nonlinear iterations, 13115 linear iterations

As you can see, the frictional contact convergence definitely takes a bigger hit from the locking correction than the frictionless contact. Frictionless contact convergence remains fine.

My contact block looks as follows
```
[Contact]
  [kontakt]
    primary = top
    secondary = rmax
    formulation = penalty
    penalty = 1e4
    model = frictionless/coulomb
    (friction_coefficient = 0.6)
    normal_smoothing_distance = 0.1
  []
[]
```

**Reply by bwspenc** [reply]
I don't know what to say about the connection between volumetric locking and poor convergence. Frictional contact problems are always tough, so it doesn't surprise me that you're in general having more trouble with friction.

It's interesting that your nonlinear convergence isn't too bad, but the linear convergence is what's really suffering. That indicates that your preconditioner isn't very effective. That could be due to a poor preconditioning method or a poor approximation of the Jacobian provided by your mechanics kernels (or a combination of the two). You'll have your best luck with a direct solver like superlu_dist and a full preconditioning matrix. I'm not sure what your settings are in your mechanics materials/kernels, but there are some options there that can help as well.

**Reply by Flolaffel** [reply]
I played with a lot of different preconditioning settings. The results posted above were obtained with PJFNK, full SMP and superlu_dist. When using fricitonless contact the best settings seem to be Newton, -ksp_type preonly and mumps. But these fail to converge with frictional contact.

Could you please provide some insight on the settings you mentioned?

**Reply by recuero** [reply]
Using volumetric locking allows for more element modes of deformation, which could theoretically make resolving friction more challenging. Even though I've never experienced that.

You could also try using the mortar approach to mechanical contact if you are not satisfied with the performance of your model with friction.

**Reply by Flolaffel** [reply]
I already tried using mortar but experienced even more problems with convergence. I followed to preconditioning guide on the "mortar performance" documentation site.

**Comment by recuero** [comment, marked as answer]
I'd recommend following the setup of existing frictional cases in the repository.
